src/Processors/Transforms/Streaming/GlobalAggregatingTransform.cpp

```diff
@@ -40,35 +41,58 @@ GlobalAggregatingTransform::GlobalAggregatingTransform(
     if (unlikely(params->params.overflow_row))
         throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Overflow row processing is not implemented in global aggregation");
 
-    /// Need extra retracted data
     if (params->emit_changelog)
     {
         if (params->emit_version)
             throw Exception(ErrorCodes::UNSUPPORTED, "'emit_version()' is not supported in global aggregation emit changelog");
 
-        ManyRetractedDataVariants retracted_data(many_data->variants.size());
-        for (auto & elem : retracted_data)
-            elem = std::make_shared<AggregatedDataVariants>();
-
+        bool retract_enabled = false;
         many_data->setField(
-            {std::move(retracted_data),
+            {retract_enabled,
              /// Field serializer
-             [this](const std::any & field, WriteBuffer & wb, VersionType) {
-                 const auto & data = std::any_cast<const ManyRetractedDataVariants &>(field);
-                 DB::writeIntBinary(data.size(), wb);
-                 for (const auto & elem : data)
-                     params->aggregator.checkpoint(*elem, wb);
+             [](const std::any & field, WriteBuffer & wb, [[maybe_unused]] VersionType version) {
+                 assert(version >= IMPL_V2_MIN_VERSION);
+                 DB::writeBoolText(std::any_cast<bool>(field), wb);
              },
              /// Field deserializer
-             [this](std::any & field, ReadBuffer & rb, VersionType) {
-                 auto & data = std::any_cast<ManyRetractedDataVariants &>(field);
-                 size_t num;
-                 DB::readIntBinary(num, rb);
-                 data.resize(num);
-                 for (auto & elem : data)
+             [this](std::any & field, ReadBuffer & rb, VersionType version) {
+                 if (version >= IMPL_V2_MIN_VERSION)
+                 {
+                     DB::readBoolText(std::any_cast<bool &>(field), rb);
```

Comment:
i think we shall use `std::variant` for field instead of std::any. `std::variant` give us a list of possibilities which are essential to serde; however std::any all kinds of possibilities.

Reply:
I don't think "Any -> Variant" makes sense, just goes from unrestricted to semi-restricted. This is no simplification either for implementation or understanding, and different implementations still need to know what they are and what they do. For `std::variant`, if there is new content, additional type declarations need to be added.

Reply:
Yes, this is the purpose to explicitly list what it is possible contains instead of a `void *`. We like to make things explicit especially for serde. When it can contain a different type like `std::string`, making it explicit has benefits we are not breaking something

Reply:
Let's defer std::any vs std::variant in future PR